colexecdisk: create disk-backed operators lazily in diskSpiller

yuzefovich opened the pull request:
This PR refactors the disk spilling machinery in the vectorized engine to create the disk-backed operator lazily, when the spilling to disk occurs, rather than eagerly as it is done right now. The main motivation is that in many cases we won't need the disk-backed operator at all, so this reduced the allocations on the hot path. Some care had to be taken to ensure that the construction - which is a planning-time operation - works when performed during the execution time. See each commit for details.

Fixes: #137323.
Release note: None

yuzefovich committed:
colexecdisk: create disk-backed operators lazily in diskSpiller

This commit makes the creation of the disk-backed operator lazy,
whenever the diskSpiller spills to disk for the first time throughout
its lifetime. This allows us avoid allocations in the common case when
we don't spill to disk.

This required adjustment for some tests that verify that the expected
closers are accumulated. In some cases we know exactly how many
closers will be added if the operator tree is forced to spill to disk
(external sort, external distinct, disk-backed hash group join) whereas
in others (external hash join and external hash aggregator) this number
is not easy to calculate. In order to have some coverage for the latter
case this commit introduces some sanity checks to ensure that at least
in some test cases the number of added closers seems reasonable.

Note that this change is not without its risks. In particular, the
disk-backed operator constructor must execute correctly when delayed.
The function captures references to miscellaneous state, which - if
mutated - can lead to problems. One such problem was the capture of
`result.ColumnTypes` when creating the external distinct, and this
commit applies a fix of having a separate reference to the input schema.
All constructor functions have been audited to ensure that no arguments
being captured would be modified later. The constructors also capture
some other functions (most commonly around the monitor registry, the
closer registry, and the constructor for disk-backed sort), but those
should be safe.

An additional complication is that the delayed constructor functions can
now run concurrently (if we have concurrency within the flow, most
likely due to the plan being distributed). To account for that the
monitor and the closer registries have been adjusted to be
concurrency-safe. Note that we could have made the mutex protection
optional (i.e. "enable" mutex protection only when we create the disk
spiller), but for safety we decided to add unconditional protection. If
mutex operations show up in the profiles we'll adjust this.

Also, note that the disk-backed operator chain will now be excluded
from EXPLAIN (VEC, VERBOSE). This seems ok.

Release note: None

yuzefovich committed:
colexec: refactor how we track closers

This commit introduces `colexecargs.CloserRegistry` which is now
responsible for tracking all `colexecop.Closer`s we create during the
vectorized planning. Previously, we would accumulate them in
`NewColOperatorResult.ToClose` slice to be appended to the `closers`
slice in the vectorized flow. However, the following commit will make
creation of some closers lazy, meaning that the captured
`result.ToClose` slice might have incomplete information, and this
commit goes around that issue.

This commit should be a no-op change as of right now.

Release note: None

yuzefovich committed:
colexec: clean up some tests a bit

The change in the previous commit made sure that we now close all
closers in the tests (if the closer registry is specified), so we
can clean up some logic around that in tests. The main idea is rather
than rely on the RunTests harness to try to close the necessary
operators, each test will now be responsible for doing so if it might
create a closer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137373)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: do not set StreamingMemAccount in some tests

Whenever we have the monitor registry in scope, we can just leave this
field unset, and an account will be created lazily via the monitor
registry. This allows for some code cleanup but also fixes an issue in
`verifyColOperator` where a memory account would be closed before those
owned by the monitor registry (which would be problematic in the
following commit where closers are closed right before the monitor
registry cleanup, but happened to be after this single acc closure).

Release note: None

yuzefovich committed:
colexec: fix Reset behavior of hash group joiner

Previously, the hash group joiner inherited the reset behavior from
`TwoInputInitHelper` that it embeds, which could leave the hash joiner
and the hash aggregator encapsulated in the hash group joiner in the
non-reset state. As a result, if the external hash group joiner spilled
to disk, we could get incorrect results. This is now fixed. There is no
explicit test added for this because the change in the following commit
(where `ChildCount` on the disk spiller now returns one less) will
exercise this code path (via "reusing after reset" part of the RunTests
harness). This is also an experimental feature, so there is no release
note.

Release note: None

yuzefovich commented:
[Here](https://gist.github.com/yuzefovich/c5b074f1804ddd2fc92470eeaddfd622) is the change on `BenchmarkExternalDistinct`. As expected, when we don't spill we see a noticeable reduction in allocations:
```
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.001/rows=1024/ordCols=0/type=bytes-24          150 ± 0%        66 ± 0%  -56.00%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.100/rows=1024/ordCols=0/type=bytes-24          150 ± 0%        66 ± 0%  -56.00%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.001/rows=1024/ordCols=0/type=int-24            154 ± 0%        70 ± 0%  -54.55%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.100/rows=1024/ordCols=0/type=int-24            168 ± 0%        84 ± 0%  -50.00%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.001/rows=65536/ordCols=0/type=bytes-24         657 ± 0%       573 ± 0%  -12.79%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.001/rows=65536/ordCols=0/type=int-24           677 ± 0%       593 ± 0%  -12.41%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.100/rows=65536/ordCols=0/type=bytes-24         681 ± 0%       597 ± 0%  -12.33%  (p=0.000 n=10+10)
ExternalDistinct/spilled=false/ordering=false/shuffled/newTupleProbability=0.100/rows=65536/ordCols=0/type=int-24           835 ± 0%       751 ± 0%  -10.06%  (p=0.000 n=10+10)
```

yuzefovich commented:
Hm, we might need mutex protection now for the monitor and closer registries.

yuzefovich commented:
Added concurrency-safety to both registries, should be RFAL.

mgartner reviewed (commented):
> Reviewed 14 of 14 files at r1, 30 of 30 files at r6, 16 of 16 files at r7, 8 of 8 files at r8, 1 of 1 files at r9, 20 of 21 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/colexec/colexecargs/monitor_registry.go` line 33 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OE__W-QATK4Wt-BO9x8:-OE__W-QATK4Wt-BO9x9:b-3tmwtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/monitor_registry.go#L33)):*
> ```Go
> // Note that Close and Reset are not protected.
> func (r *MonitorRegistry) SetMutex(mu *syncutil.Mutex) {
> 	if r.optionalMu == nil {
> ```

I have the same concern here—if tow operators spill to disk at the same time, this would be a race condition, wouldn't it?
___
*[`pkg/sql/colexec/colexecargs/closer_registry.go` line 27 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OE__0eO0iPHgKzsfTai:-OE__0eO0iPHgKzsfTaj:b-ti8x3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/closer_registry.go#L27)):*
> ```Go
> // Note that Close and Reset are not protected.
> func (r *CloserRegistry) SetMutex(mu *syncutil.Mutex) {
> 	if r.optionalMu == nil {
> ```

Couldn't we have a race if two goroutines simultaneously try to set this mutex?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/colexec/colexecargs/closer_registry.go` line 27 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OE__0eO0iPHgKzsfTai:-OEenPSTAnRvfN4A5vTA:b-p5t6yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/closer_registry.go#L27)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Couldn't we have a race if two goroutines simultaneously try to set this mutex?
</blockquote></details>

`SetMutex` is called during the operator planning which is done in a single goroutine - this is done in `makeConstructorConcurrencySafe` which is called whenever we create a disk-spilling operator. Only `diskBackedConstructor` function can be called during the execution, and at that point the mutex will have been set, protecting the registries.

I tried to deduplicate the code a bit with `makeConstructorConcurrencySafe` but perhaps it makes it harder to reason about - should we just explicitly call `SetMutex` right before calling disk spiller constructors?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/colexec/colexecargs/closer_registry.go` line 27 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OE__0eO0iPHgKzsfTai:-OEo0sDb-1meQsr11yky:bbdnw5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/closer_registry.go#L27)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`SetMutex` is called during the operator planning which is done in a single goroutine - this is done in `makeConstructorConcurrencySafe` which is called whenever we create a disk-spilling operator. Only `diskBackedConstructor` function can be called during the execution, and at that point the mutex will have been set, protecting the registries.

I tried to deduplicate the code a bit with `makeConstructorConcurrencySafe` but perhaps it makes it harder to reason about - should we just explicitly call `SetMutex` right before calling disk spiller constructors?
</blockquote></details>

You mean before calling the functions like`NewOneInputDiskSpiller`? I don't have a very strong opinion here, so up to you. Not sure the `makeConstructorConcurrencySafe` adds much value, but either way seems potentially easy to misuse. 

Do you think there are any valuable assertions we could add, e.g., is there a way that `AddCloser` could assert that the mutex has been set when it should be?
___
*[`pkg/sql/colexec/colexecargs/op_creation.go` line 109 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OEo0Z9a3N5PFgbagVFC:-OEo0Z9a3N5PFgbagVFD:b-tm33ik) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/op_creation.go#L109)):*
> ```Go
> func (r *NewColOperatorArgs) MakeConcurrencySafeForDiskBackedOp() {
> 	r.MonitorRegistry.SetMutex(r.RegistriesMu)
> 	r.CloserRegistry.SetMutex(r.RegistriesMu)
> ```

Do these need to be the same mutex? Would using a mutex embedded in the registries with a boolean indicating whether it should be used or not be simpler?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/colexec/colexecargs/closer_registry.go` line 27 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OE__0eO0iPHgKzsfTai:-OEpIPs4COehfDzQD0R4:b-qy2g26) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/closer_registry.go#L27)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You mean before calling the functions like`NewOneInputDiskSpiller`? I don't have a very strong opinion here, so up to you. Not sure the `makeConstructorConcurrencySafe` adds much value, but either way seems potentially easy to misuse. 

Do you think there are any valuable assertions we could add, e.g., is there a way that `AddCloser` could assert that the mutex has been set when it should be?
</blockquote></details>

I decided to remove this optionality and just make the registries concurrency-safe unconditionally. Mutex operations should take on the order of tens nanoseconds, and during the operator planning we only have a single goroutine, so there shouldn't be any contention on the hot path. In other words, let's make it safe first, and if we see the mutex overhead show up in profiles, then we'll optimize it.
___
*[`pkg/sql/colexec/colexecargs/op_creation.go` line 109 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OEo0Z9a3N5PFgbagVFC:-OEpIBkuDxUo2bvbSfFG:b-ytfpnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/op_creation.go#L109)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Do these need to be the same mutex? Would using a mutex embedded in the registries with a boolean indicating whether it should be used or not be simpler?
</blockquote></details>

The mutexes can be different. Yeah, perhaps a boolean to indicate whether the concurrency protection is enabled would be cleaner. I put this thought aside for now (see previous comment).


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Thanks for doing this!

> Reviewed 9 of 32 files at r13, 6 of 16 files at r14, 5 of 8 files at r15, 1 of 1 files at r16, 20 of 20 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-:-OG68TiLAkfbUaYDQbVF:b-omhxxd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/colexec/colexecargs/closer_registry.go` line 27 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OE__0eO0iPHgKzsfTai:-OG67LXY9xPTzU3CD8M0:bf8ptrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/closer_registry.go#L27)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I decided to remove this optionality and just make the registries concurrency-safe unconditionally. Mutex operations should take on the order of tens nanoseconds, and during the operator planning we only have a single goroutine, so there shouldn't be any contention on the hot path. In other words, let's make it safe first, and if we see the mutex overhead show up in profiles, then we'll optimize it.
</blockquote></details>

SGTM :+1: 
___
*[`pkg/sql/colexec/colexecargs/op_creation.go` line 109 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-OEo0Z9a3N5PFgbagVFC:-OG67kNiAirTenk6FA2i:bmeyrwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcdf26436391eef52fcd235b26b2972914b17d39/pkg/sql/colexec/colexecargs/op_creation.go#L109)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The mutexes can be different. Yeah, perhaps a boolean to indicate whether the concurrency protection is enabled would be cleaner. I put this thought aside for now (see previous comment).
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->

mgartner commented:
@yuzefovich Did you run any benchmarks with this PR? I can run some quick sysbench microbenchmarks if not.

yuzefovich reviewed (commented):
Yeah, I included some microbenchmarks in [this](https://github.com/cockroachdb/cockroach/pull/137373#issuecomment-2542582879) comment. I also just reran the benchmarks with the latest version, the results show an improvement https://gist.github.com/yuzefovich/c5b074f1804ddd2fc92470eeaddfd622.

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137373#-:-OG6FMcQ-Af5zJqZ61B5:b2dp908)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->
